File: 2025/07 July/09 Sum of subarray minimum/main.cpp
// Link: https://www.geeksforgeeks.org/problems/sum-of-subarray-minimum/1

#include <iostream>
#include <vector>
#include <stack>
#include <numeric>
using namespace std;

class Solution
{
public:
    int sumSubMins(vector<int> &arr)
    {
        int n = arr.size();
        vector<int> dp(n, 0);
        vector<int> right(n);
        vector<int> stack;

        for (int i = 0; i < n; i++)
            right[i] = i;

        for (int i = 0; i < n; i++)
        {
            while (!stack.empty() && arr[i] < arr[stack.back()])
            {
                right[stack.back()] = i;
                stack.pop_back();
            }
            stack.push_back(i);
        }

        dp[n - 1] = arr[n - 1];
        for (int i = n - 2; i >= 0; i--)
        {
            int r = right[i];
            if (r == i)
            {
                dp[i] = (n - i) * arr[i];
            }
            else
            {
                dp[i] = (r - i) * arr[i] + dp[r];
            }
        }
        return accumulate(dp.begin(), dp.end(), 0);
    }
};

int main()
{
    Solution sol;
    vector<int> arr = {3, 1, 2, 4};
    int result = sol.sumSubMins(arr);
    cout << "Total sum of the minimum elements of every possible subarrays is " << result << endl;

    arr = {71, 55, 82, 55};
    result = sol.sumSubMins(arr);
    cout << "Total sum of the minimum elements of every possible subarrays is " << result << endl;
    return 0;
}

File: 2025/07 July/09 Sum of subarray minimum/main.py
# Link: https://www.geeksforgeeks.org/problems/sum-of-subarray-minimum/1


class Solution:
    def sumSubMins(self, arr):
        n = len(arr)
        dp = [0] * n
        right = [i for i in range(n)]
        stack = []

        for i in range(n):
            while stack and arr[i] < arr[stack[-1]]:
                right[stack.pop()] = i
            stack.append(i)

        dp[n - 1] = arr[n - 1]
        for i in range(n - 2, -1, -1):
            r = right[i]
            if r == i:
                dp[i] = (n - i) * arr[i]
            else:
                dp[i] = (r - i) * arr[i] + dp[r]

        return sum(dp)


if __name__ == "__main__":
    sol = Solution()
    arr = [3, 1, 2, 4]
    result = sol.sumSubMins(arr)
    print(f"Total sum of the minimum elements of every possible subarrays is {result}")

    arr = [71, 55, 82, 55]
    result = sol.sumSubMins(arr)
    print(f"Total sum of the minimum elements of every possible subarrays is {result}")
